src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,153 @@
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> current = tail;
+        Node<T> newNode = new Node<>(current, value,null);
+        tail = newNode;
+        if (current == null) {
+            head = newNode;
+        } else {
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = findNodeByIndex(index);
+            Node<T> prev = current.prev;
+            Node<T> newNode = new Node<>(prev, value, current);
+            current.prev = newNode;
+            if (prev == null) {
+                head = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            size++;
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < (size >> 1)) {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            return current;
+        } else {
+            Node<T> current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+            return current;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+        Node<T> x = findNodeByIndex(index);
+        T oldValue = x.value;
+        x.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+        Node<T> current = findNodeByIndex(index);
+        removeNode(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current.next != null) {
+            if ((current.value == object) || (current.value != null && current.value.equals(object))) {
+                break;
+            }
+            current = current.next;
+        }
+        if (Objects.equals(current.value, object)) {
+            removeNode(current);
+            return true;
+        }
         return false;
     }
 
+    private void removeNode(Node<T> current) {
```

Comment:
```suggestion
    private void unlink(Node<T> current) {
```
removeNode when it is called from remove method is really not informative name